meos/src/point/tpoint_rtree.c

```diff
@@ -0,0 +1,663 @@
+      i--;
+    }
+  }
+
+  // Make sure that both left and right nodes have at least
+  // MINITEMS by moving data into underflowed nodes.
+  if (node -> count < MINITEMS) {
+    // reverse sort by min axis
+    node_sort_axis(rtree, right, largest_axis, false);
+    do {
+      node_move_box_at_index_into(right, right -> count, node);
+    } while (node -> count < MINITEMS);
+  } else if (right -> count < MINITEMS) {
+    // reverse sort by max axis
+    node_sort_axis(rtree, node, largest_axis, true);
+    do {
+      node_move_box_at_index_into(node, node -> count - 1, right);
+    } while (right -> count < MINITEMS);
+  }
+  if (node -> kind == RTREE_INNER_NODE) {
+    node_sort_axis(rtree, node, 0, false);
+    node_sort_axis(rtree, right, 0, false);
+  }
+  * right_out = right;
+}
+
+/**
+ * @brief Inserts an STBox into an RTree node and handles node splitting if necessary.
+ * @details This function inserts a new STBox into an RTree node. If the node is a leaf and 
+ * already contains the maximum number of items (`MAXITEMS`), it sets the `split` flag to `true`
+ * to indicate that the node needs to be split. For non-leaf nodes, the function determines 
+ * the appropriate child node for insertion and recursively inserts the STBox. If splitting occurs, 
+ * the function handles the split and updates the parent node's bounding boxes.
+ * @param[in] rtree Pointer to the RTree structure that provides axis value retrieval and node splitting functions.
+ * @param[in] old_box Pointer to the STBox that is being replaced or added.
+ * @param[in] node Pointer to the RTreeNode structure where the STBox is being inserted.
+ * @param[in] new_box Pointer to the STBox to be inserted.
+ * @param[in] id Identifier associated with the new STBox (used only for leaf nodes).
+ * @param[out] split Pointer to a boolean flag that indicates if the node was split during insertion.
+ */
+static void
+node_insert(RTree * rtree, STBox * old_box, RTreeNode * node,
+  STBox * new_box, int id, bool * split) {
+  if (node -> kind == RTREE_INNER_NODE_NO) {
+    if (node -> count == MAXITEMS) {
+      * split = true;
+      return;
+    }
+    int index = node -> count;
+    node -> boxes[index] = * new_box;
+    node -> ids[index] = id;
+    node -> count++;
+    * split = false;
+    return;
+  }
+  int insertion_node = node_choose(rtree, new_box, node);
+  node_insert(rtree, & node -> boxes[insertion_node], (RTreeNode * ) node -> nodes[insertion_node], new_box, id, split);
+  if (! * split) {
+    stbox_expand(new_box, & node -> boxes[insertion_node]);
+    * split = false;
+    return;
+  }
+  if (node -> count == MAXITEMS) {
+    * split = true;
+    return;
+  }
+  RTreeNode * right;
+  node_split(rtree, node -> nodes[insertion_node], & node -> boxes[insertion_node], & right);
+  node -> boxes[insertion_node] = * node_box_calculate(node -> nodes[insertion_node]);
+  node -> boxes[node -> count] = * node_box_calculate(right);
```

Comment:
You probaby want `node_box_calculate` to return a stack-allocated `STBox` instead of a pointed to a heap-allocated box. Indeed, currently you are copying its value and you are never freeing the returned box so you probably don't want to allocate it in the first place.

Reply:
done!